File: data-structures/ring-buffer/tests/ring-buffer.test.ts
import assert from "node:assert/strict";
import { MyRingBuffer } from "../implementations/ring-buffer";

function runTest(name: string, execute: () => void): void {
  try {
    execute();
    console.log(`PASS ${name}`);
  } catch (error) {
    console.error(`FAIL ${name}`);
    throw error;
  }
}

runTest("simple: new ring buffer starts empty", () => {
  const buffer = new MyRingBuffer<number>(3);

  assert.equal(buffer.length(), 0);
  assert.equal(buffer.capacity(), 3);
  assert.equal(buffer.isEmpty(), true);
  assert.equal(buffer.isFull(), false);
  assert.equal(buffer.front(), undefined);
  assert.equal(buffer.rear(), undefined);
  assert.deepEqual(buffer.toArray(), []);
});

runTest("simple: enqueue fills the buffer and exposes front/rear", () => {
  const buffer = new MyRingBuffer<number>(3);

  assert.equal(buffer.enqueue(10), true);
  assert.equal(buffer.enqueue(20), true);
  assert.equal(buffer.enqueue(30), true);

  assert.equal(buffer.length(), 3);
  assert.equal(buffer.isFull(), true);
  assert.equal(buffer.front(), 10);
  assert.equal(buffer.rear(), 30);
  assert.deepEqual(buffer.toArray(), [10, 20, 30]);
});

runTest("simple: enqueue on a full buffer is rejected and leaves it untouched", () => {
  const buffer = new MyRingBuffer<number>(2);

  buffer.enqueue(1);
  buffer.enqueue(2);

  assert.equal(buffer.enqueue(3), false);
  assert.equal(buffer.length(), 2);
  assert.deepEqual(buffer.toArray(), [1, 2]);
});

runTest("simple: dequeue returns items in FIFO order and frees a slot", () => {
  const buffer = new MyRingBuffer<string>(2);

  buffer.enqueue("a");
  buffer.enqueue("b");

  assert.equal(buffer.dequeue(), "a");
  assert.equal(buffer.isFull(), false);
  assert.equal(buffer.front(), "b");
  assert.equal(buffer.dequeue(), "b");
  assert.equal(buffer.dequeue(), undefined);
});

runTest("simple: clear resets the buffer", () => {
  const buffer = new MyRingBuffer<number>(2);

  buffer.enqueue(1);
  buffer.enqueue(2);
  buffer.clear();

  assert.equal(buffer.length(), 0);
  assert.equal(buffer.isEmpty(), true);
  assert.equal(buffer.isFull(), false);
  assert.equal(buffer.dequeue(), undefined);
});

runTest("mildly absurd: repeated wraparound keeps contents correct across many cycles", () => {
  const capacity = 4;
  const buffer = new MyRingBuffer<number>(capacity);
  const reference: number[] = [];

  for (let index = 1; index <= 60; index += 1) {
    const shouldEnqueue = reference.length < capacity;
    const enqueued = buffer.enqueue(index);

    assert.equal(enqueued, shouldEnqueue);
    if (shouldEnqueue) {
      reference.push(index);
    }

    if (index % 3 === 0 && reference.length > 0) {
      buffer.dequeue();
      reference.shift();
    }
  }

  assert.deepEqual(buffer.toArray(), reference);
  assert.equal(buffer.capacity(), capacity);
});

runTest("absurd: deterministic operation sequence matches a reference model", () => {
  const capacity = 5;
  const buffer = new MyRingBuffer<number>(capacity);
  const reference: number[] = [];

  for (let index = 1; index <= 200; index += 1) {
    const shouldEnqueue = reference.length < capacity;
    const enqueued = buffer.enqueue(index);

    assert.equal(enqueued, shouldEnqueue);
    if (shouldEnqueue) {
      reference.push(index);
    }

    if (index % 3 === 0 && reference.length > 0) {
      buffer.dequeue();
      reference.shift();
    }
    if (index % 7 === 0 && reference.length > 0) {
      buffer.dequeue();
      reference.shift();
    }
  }

  assert.deepEqual(buffer.toArray(), reference);
  assert.equal(buffer.length(), reference.length);
  assert.equal(buffer.front(), reference[0]);
  assert.equal(buffer.rear(), reference[reference.length - 1]);
  assert.equal(buffer.capacity(), capacity);
});

File: data-structures/ring-buffer/tests/ring_buffer_test.go
package tests

import (
	"reflect"
	"testing"

	"dsa/data-structures/ring-buffer/implementations"
)

func TestRingBufferSimpleNewBufferStartsEmpty(t *testing.T) {
	buffer := implementations.NewMyRingBuffer[int](3)

	if got := buffer.Length(); got != 0 {
		t.Fatalf("expected length 0, got %d", got)
	}

	if got := buffer.Capacity(); got != 3 {
		t.Fatalf("expected capacity 3, got %d", got)
	}

	if !buffer.IsEmpty() {
		t.Fatal("expected buffer to be empty")
	}

	if buffer.IsFull() {
		t.Fatal("expected buffer to not be full")
	}

	if got := buffer.ToSlice(); len(got) != 0 {
		t.Fatalf("expected empty slice, got %v", got)
	}
}

func TestRingBufferSimpleEnqueueFillsBufferAndExposesFrontRear(t *testing.T) {
	buffer := implementations.NewMyRingBuffer[int](3)

	if !buffer.Enqueue(10) {
		t.Fatal("expected enqueue of 10 to succeed")
	}
	if !buffer.Enqueue(20) {
		t.Fatal("expected enqueue of 20 to succeed")
	}
	if !buffer.Enqueue(30) {
		t.Fatal("expected enqueue of 30 to succeed")
	}

	if got := buffer.Length(); got != 3 {
		t.Fatalf("expected length 3, got %d", got)
	}

	if !buffer.IsFull() {
		t.Fatal("expected buffer to be full")
	}

	if got := buffer.Front(); got != 10 {
		t.Fatalf("expected front 10, got %d", got)
	}

	if got := buffer.Rear(); got != 30 {
		t.Fatalf("expected rear 30, got %d", got)
	}

	if got := buffer.ToSlice(); !reflect.DeepEqual(got, []int{10, 20, 30}) {
		t.Fatalf("expected [10 20 30], got %v", got)
	}
}

func TestRingBufferSimpleEnqueueOnFullBufferIsRejected(t *testing.T) {
	buffer := implementations.NewMyRingBuffer[int](2)

	buffer.Enqueue(1)
	buffer.Enqueue(2)

	if buffer.Enqueue(3) {
		t.Fatal("expected enqueue on a full buffer to fail")
	}

	if got := buffer.Length(); got != 2 {
		t.Fatalf("expected length 2, got %d", got)
	}

	if got := buffer.ToSlice(); !reflect.DeepEqual(got, []int{1, 2}) {
		t.Fatalf("expected [1 2], got %v", got)
	}
}

func TestRingBufferSimpleDequeueReturnsItemsInFIFOOrder(t *testing.T) {
	buffer := implementations.NewMyRingBuffer[string](2)

	buffer.Enqueue("a")
	buffer.Enqueue("b")

	if got := buffer.Dequeue(); got != "a" {
		t.Fatalf("expected a, got %s", got)
	}

	if buffer.IsFull() {
		t.Fatal("expected buffer to not be full after dequeue")
	}

	if got := buffer.Front(); got != "b" {
		t.Fatalf("expected front b, got %s", got)
	}
}

func TestRingBufferSimpleClearResetsTheBuffer(t *testing.T) {
	buffer := implementations.NewMyRingBuffer[int](2)

	buffer.Enqueue(1)
	buffer.Enqueue(2)
	buffer.Clear()

	if got := buffer.Length(); got != 0 {
		t.Fatalf("expected length 0, got %d", got)
	}

	if !buffer.IsEmpty() {
		t.Fatal("expected buffer to be empty")
	}
}

func TestRingBufferMildlyAbsurdRepeatedWraparound(t *testing.T) {
	capacity := 4
	buffer := implementations.NewMyRingBuffer[int](capacity)
	reference := make([]int, 0)

	for index := 1; index <= 60; index++ {
		shouldEnqueue := len(reference) < capacity
		enqueued := buffer.Enqueue(index)

		if enqueued != shouldEnqueue {
			t.Fatalf("index %d: expected enqueue success=%v, got %v", index, shouldEnqueue, enqueued)
		}
		if shouldEnqueue {
			reference = append(reference, index)
		}

		if index%3 == 0 && len(reference) > 0 {
			buffer.Dequeue()
			reference = reference[1:]
		}
	}

	if got := buffer.ToSlice(); !reflect.DeepEqual(got, reference) {
		t.Fatalf("expected %v, got %v", reference, got)
	}

	if got := buffer.Capacity(); got != capacity {
		t.Fatalf("expected capacity to stay %d, got %d", capacity, got)
	}
}

func TestRingBufferAbsurdDeterministicOperationSequenceMatchesReferenceModel(t *testing.T) {
	capacity := 5
	buffer := implementations.NewMyRingBuffer[int](capacity)
	reference := make([]int, 0)

	for index := 1; index <= 200; index++ {
		shouldEnqueue := len(reference) < capacity
		enqueued := buffer.Enqueue(index)

		if enqueued != shouldEnqueue {
			t.Fatalf("index %d: expected enqueue success=%v, got %v", index, shouldEnqueue, enqueued)
		}
		if shouldEnqueue {
			reference = append(reference, index)
		}

		if index%3 == 0 && len(reference) > 0 {
			buffer.Dequeue()
			reference = reference[1:]
		}
		if index%7 == 0 && len(reference) > 0 {
			buffer.Dequeue()
			reference = reference[1:]
		}
	}

	if got := buffer.ToSlice(); !reflect.DeepEqual(got, reference) {
		t.Fatalf("expected %v, got %v", reference, got)
	}

	if got := buffer.Length(); got != len(reference) {
		t.Fatalf("expected length %d, got %d", len(reference), got)
	}
}
